### PR TITLE
draft

### PR DESCRIPTION
**Pull request by Dean-Kim0507**
#3 issue: Database Redesign
The chatting room will allow chatting with groups of 3 or more users. The host can open the chat room and invite people by searching username. The host set the limit number of members for the conversation room so if it’s two this is a 1:1 room likewise the previous one. The chat room type will be classified by the number of limit members attribute.

### Database
- Add a table “User_Conversation” to make a M:N relation between User and Conversation.
To allow conversations with groups of 3 or more users, the user model needs to have many conversations and the conversation model also needed to have many users. A Many-to-Many relationship has to be broken down into intermediary tables because it makes a lot of redundancy. Bridge entity is a solution for implementing the M:N relationship and which are connecting each other by using a foreign key. 
- Delete User1, User2 Column.
These attributes are no longer needed, participants can be searched by using the user_conversation table.

ERD with colored entities (UML notation)

### Server
- a find conversation method in the server is needed to change to using a user_conversation table rather than using user1 or user2 attributes.
- Creating a conversation room will be joined to create data in the bridge table(user_conversation).
- new messages should emit to users in the chat room, not all users for security.

### Frontend
- Make a UI to invite other users and set the number of participants. 

## 2. If our app were already deployed, what steps would we need to take to move to this new feature without disrupting service for current users?

Production redeployment involves deploying a new version of the application and an older version of the same application. Clients already connected to the application during the redeployment continue to use the older version of the application until they complete their work. 
By using canary or blue-green deployment, user traffic gradually transfers from the previous version to the new version, and the risk can be reduced. 
Huge changes can make bugs we didn’t expect so changes could happen gradually.
1.	Planning and designing a release.
2.	Software build
3.	Test to the stakeholders and get a feedbacks
4.	Announce the updating date, time and new features to the users.
5.	Test a new version of the application
6.	Deployment
7.	Issue reporting and collection.

### CODE REVIEW

File: server/db/models/index.js [hunk 1]
```diff
@@ -4,12 +4,23 @@ const Message = require("./message");
...
-Conversation.belongsTo(User, { as: "user2" });
+User.belongsToMany(Conversation,{
+  through: 'user_conversations',
+  foreignKey: 'user_id'
+})
+
+Conversation.belongsToMany(User,{
+  through: 'user_conversations',
+  foreignKey: 'convo_id'
+})
```
Comment (johnsciutto) [inline server/db/models/index.js:15]: 👍🏻

File: server/db/models/index.js [hunk 2]
```diff
@@ -4,12 +4,23 @@ const Message = require("./message");
...
+
 Message.belongsTo(Conversation);
 Conversation.hasMany(Message);
 
+//Full features are not implemented, I'll leave these just for passing a test.
+Conversation.belongsTo(User, { as: "user1" });
+Conversation.belongsTo(User, { as: "user2" });
```
Comment (johnsciutto) [inline server/db/models/index.js:33]: This PR is not intended to be merged in, so it's OK if the test is not passing 🙂

**Review by johnsciutto — COMMENTED** [review]
Hi @Dean-Kim0507,
Nice work on this PR 👍🏻 🙂 
I have a couple of additional questions:

- What changes would you make to the API routes (if any) to implement this new feature?
- If you were working on this feature with a team of several developers, how would you divide the tasks among your teammates to minimize merge-conflicts?

Let us know when the questions are answered by using the "/hatchy pr" command on Slack again and another reviewer will come by to give this PR another look and give you approval 👍🏻

**Comment by Dean-Kim0507** [comment]
Additional Question
## What changes would you make to the API routes (if any) to implement this new feature?
### Conversation
-	createChatRoom (post) - This function is done by the message API (if convesationId doesn’t exist in the database, make a conversation and return message info with new conversationId). To make a group meeting, additional parameters are needed e.g. inviting members, setting limit numbers, hostId, and so on.  Making a chatroom will be an independent API to create a conversation room with additional information.
-	levaeChatroom (put) - If one of the users leaves the chat room, the API updates the user info in the user and conversation database.
-	InviteUsers (put) – Current user can invite new users. This API updates the current user info in the database.
-	editChatroom (put) – The host can change the chatting room’s information.
### Message
I added the additional change Message mode. The message model should make M:N relationship with User mode to track the read status of each member.
-	readStatusChange (put) – Change the read status when the user click the chat room.
-	postMessage (post) – Read status will be true when active conversationId of the user is the same as this conversation room id(which means that user is now looking at this conversation room).

ERD with colored entities (UML notation)
## If you were working on this feature with a team of several developers, how would you divide the tasks among your teammates to minimize merge-conflicts?
-	To minimize merge-conflicts, the best way is to split the work between developers into file units. E.g. a developer works on conversation AIPs, another developer work on message APIs.
-	If two developers needed to work on the same file, use a new file in preference to an existing one.
-	Push and pull changes are often as you can.

**Review by FilipeGorodscy — APPROVED** [review]
Approved, but please don't merge since it's only a draft PR :)

**Pull request closed without merging**